### PR TITLE
Migrate authn filter to jBallerina

### PR DESCRIPTION
**Pull request by Rajith90**
### Purpose
With jBallerina authn filter is provided by b7a layer. So in micro gateway level we don't need an authentication filter. We need a pre authentication filter to set the additional information prior to the authentication.

### Issues

Fixes #717 

### Automation tests
 - Unit tests added: Yes/No
 - Integration tests added: Yes/No

### Tested environments

Not Tested

---
#### Maintainers: Check before merge
- [x] Assigned 'Type' label
- [x] Assigned the project
- [x] Validated respective github issues
- [x] Assigned milestone to the github issue(s)

### CODE REVIEW

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal [hunk 1]
```diff
@@ -0,0 +1,157 @@
+// Copyright (c)  WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]: License year?
```suggestion
// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    public function filterRequest(http:Caller caller, http:Request request, http:FilterContext context)
+                        returns boolean {
+        //Setting UUID
+        int startingTime = getCurrentTime();
+        context.attributes[REQUEST_TIME] = startingTime;
+        checkOrSetMessageID(context);
+        setHostHeaderToFilterContext(request, context);
+        setLatency(startingTime, context, SECURITY_LATENCY_AUTHN);
+        return doAuthnFilterRequest(caller, request, context);
+
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]:
```suggestion
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+    string authHeader = "";
+    string? authCookie = "";
+    string|error extractedToken = "";
+    string authHeaderName = getAuthorizationHeader(invocationContext);
+    invocationContext.attributes[AUTH_HEADER] = authHeaderName;
+    string[] authProvidersIds = getAuthProviders(context.getServiceName());
+
+    if (request.hasHeader(authHeaderName)) {
+            authHeader = request.getHeader(authHeaderName);
+    } else if (request.hasHeader(COOKIE_HEADER)){
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]:
```suggestion
    } else if (request.hasHeader(COOKIE_HEADER)) {
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    
+    string providerId;
+    if (!isCookie) {
+        providerId = getAuthenticationProviderType(authHeader);
+    } else {
+        providerId = getAuthenticationProviderTypeWithCookie(authHeader);
+    }
+    boolean canHandleAuthentication = false;
+    foreach string provider in authProvidersIds {
+        if(provider == providerId) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]:
```suggestion
        if (provider == providerId) {
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            canHandleAuthentication = true;
+        }
+    }
+
+    if(!canHandleAuthentication) {
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]:
```suggestion
    if (!canHandleAuthentication) {
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        printDebug(KEY_AUTHN_FILTER, "Setting the backed up auth header value to the header: " + AUTH_HEADER);
+        request.removeHeader(TEMP_AUTH_HEADER);
+        printDebug(KEY_AUTHN_FILTER, "Removed header : " + TEMP_AUTH_HEADER + " from the request");
+    }
+}
+
+function getAuthCookieIfPresent(http:Request request) returns string? {
+    //get required cookie as config value
+    string? authCookie = ();
+        if (request.hasHeader(COOKIE_HEADER)){
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]: if block is not indented properly also,
```suggestion
        if (request.hasHeader(COOKIE_HEADER)) {
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/cookie_handler.bal [hunk 7]
```diff
@@ -0,0 +1,68 @@
+// Copyright (c)  WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/cookie_handler.bal]: License year?
```suggestion
// Copyright (c) 2019 WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/cookie_handler.bal [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+
+
+public type CookieAuthHandler object {
+
+    # Checks if the request can be authenticated with the Cookie Auth header.
+    #
+    # + req - The `Request` instance.
+    # + return - Returns `true` if can be authenticated. Else, returns `false`.
+    public function canProcess(http:Request req) returns @tainted boolean {
+        if (req.hasHeader(COOKIE_HEADER)){
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/cookie_handler.bal]:
```suggestion
        if (req.hasHeader(COOKIE_HEADER)) {
```

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal [hunk 9]
```diff
@@ -0,0 +1,157 @@
...
+
+    if(!canHandleAuthentication) {
+        setErrorMessageToFilterContext(context, API_AUTH_PROVIDER_INVALID);
+        sendErrorResponse(caller, request, context);
+        return false;
+    }
+    // if auth providers are there, use those to authenticate
+    
+        //TODO: Move this to post authentication handler
+        //checkAndRemoveAuthHeaders(request, authHeaderName);
```
Comment (praminda) [inline components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]: Is this `TODO:` still something we need to work on or resolved?

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal (hunk 1 above)
Comment (Rajith90) [reply components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]: These are not new files, I have renamed and moved them

File: components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal (hunk 9 above)
Comment (Rajith90) [reply components/micro-gateway-core/src/main/ballerina/src/gateway/filters/pre_authn_filter.bal]: We need to work on this, not resolved

File: components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/cookie_handler.bal (hunk 7 above)
Comment (Rajith90) [reply components/micro-gateway-core/src/main/ballerina/src/gateway/handlers/cookie_handler.bal]: These are not new files, I have renamed and moved them

**Pull request merged**